lib/ret/room_object.ex

```diff
@@ -0,0 +1,73 @@
+defmodule Ret.RoomObject do
+  use Ecto.Schema
+  import Ecto.Changeset
+  import Ecto.Query
+
+  alias Ret.{EncryptedField, Hub, RoomObject, Repo}
+  @schema_prefix "ret0"
+  @primary_key {:room_object_id, :id, autogenerate: true}
+
+  schema "room_objects" do
+    field(:room_object_sid, :string)
+    field(:gltf_node, EncryptedField)
+
+    belongs_to(:hub, Hub, references: :hub_id)
+
+    timestamps()
+  end
+
+  def perform_pin!(%Hub{hub_id: hub_id} = hub, %{room_object_sid: room_object_sid} = attrs) do
+    attrs = attrs |> Map.put(:gltf_node, attrs |> Map.get(:gltf_node) |> Poison.encode!())
+
+    room_object =
+      RoomObject
+      |> where([t], t.hub_id == ^hub_id and t.room_object_sid == ^room_object_sid)
+      |> preload(:hub)
+      |> Repo.one()
+
+    changeset(room_object || %RoomObject{}, hub, attrs) |> Repo.insert_or_update!()
+    Cachex.expire(:room_gltf, hub_id, -1)
+  end
+
+  def perform_unpin(%Hub{hub_id: hub_id}, room_object_sid) do
+    RoomObject
+    |> where([t], t.hub_id == ^hub_id and t.room_object_sid == ^room_object_sid)
+    |> Repo.delete_all()
+
+    Cachex.expire(:room_gltf, hub_id, -1)
+  end
+
+  def gltf_for_hub_id(hub_id) do
+    nodes =
+      RoomObject
+      |> where([t], t.hub_id == ^hub_id)
+      |> Repo.all()
+      |> Enum.map(&(&1.gltf_node |> Poison.decode!()))
+
+    node_indices =
+      if length(nodes) == 0 do
+        []
+      else
+        0..((nodes |> length) - 1) |> Enum.to_list()
+      end
+
+    gltf =
+      %{
+        asset: %{version: "2.0", generator: "reticulum"},
+        scenes: [%{nodes: node_indices, name: "Room Objects"}],
+        nodes: nodes,
+        extensionsUsed: ["HUBS_components"]
+      }
+      |> Poison.encode!()
+
+    {:commit, gltf}
+  end
+
+  defp changeset(%RoomObject{} = room_object, %Hub{} = hub, attrs) do
+    room_object
+    |> cast(attrs, [:room_object_sid, :gltf_node])
+    |> unique_constraint(:room_object_sid)
+    |> unique_constraint(:hub_id)
```

Comment:
Reading up on unique constraints, should we be using the combined constraint name here? i.e. `unique_constraint(:room_objects_room_object_sid_hub_id_index)`
https://hexdocs.pm/ecto/Ecto.Changeset.html#unique_constraint/3-complex-constraints

Reply:
interesting, TIL. i'm not sure what the effect of adding this to the changeset actually is, so I'll look into that too.